sympy/printing/latex.py

```diff
@@ -30,6 +30,7 @@
 import re
 
 if TYPE_CHECKING:
+    from sympy.tensor.array import NDimArray
     from sympy.vector.basisdependent import BasisDependent
```

Comment:
What is the purpose of this?

I think so far `if TYPE_CHECKING` has only been used to prevent circular imports.

Reply:
We were hoping to reduce the load of ``import sympy``, but the initial solution seemed to cause some circular imports. This was one trivial case to fix.

Reply:
> We were hoping to reduce the load of `import sympy`

Have you done any timing measurements?

Does this particular change give any noticeable improvement or is there a reason to think that a few changes like this might lead to an improvement?

The tensor module is imported during `import sympy` because `NDimArray` is imported up to top-level so moving imports around here will not affect the fact that the module always gets imported. This is an unfortunate consequence of wanting everything to be available like `from sympy import NDimArray` and it is not something that can be changed without breaking backwards compatibility.

Reply:
I haven't got any timings for this specific case, no. The improvement for this individual change is likely to be negligible. 

I do believe that other changes (for which this change is a requirement) will have an impact (based on the changes I have locally). I can't guarantee there are no regressions in my local changes but I have reduced the total time for ``import sympy``. But, if I verify the other changes work well, it can be a discussion for another time.

The reason for raising this PR now, and alone was more because it was such a trivial change, and IMO makes sense. Having this import within the ``TYPE_CHECKING`` block makes more sense than the way it was previously (IMO). It should not affect any other code (unless someone imports ``NDimArray`` from ``sympy/printing/latex.py``, unlikely). This PR should be viewed independently to improving the runtime of ``import numpy``.

Reply:
> I do believe that other changes (for which this change is a requirement) will have an impact

How is this change a requirement for anything else?

Note that there is a tension the other way: some downstream things want to be able to access the type hints at runtime which breaks if `if TYPE_CHECKING` is used e.g. gh-23216.

Reply:
> How is this change a requirement for anything else?

This has detoured. I am not saying this change is a requirement for anything. I'm not saying it should be merged if people disagree (this is the purpose of a review process, right?). However, there is a change I would like to use in our codebase, to speed up ``import sympy``. In order to fully implement that change I will need to solve some circular imports, this import causes one of these circular imports. But, lets forget that this change is a prerequisite for anything else from here on in, it is not at all relevant (and becoming a distraction) to why I chose to raise a PR now.


The purpose of this change is that this import solely exists to type hint, it is not used at runtime. For this reason, there exists a ``typing.TYPE_CHECKING`` boolean. This change is to ensure this import is not executed at runtime. I also verified this with examples in the sympy codebase, and the issue you mentioned, #23216 was an example I used. A further example that I used is in this same file: 

``from sympy.vector.basisdependent import BasisDependent``

 ``BasisDependent`` is used in exactly the same way as ``NDimArray``, and is protected by the ``typing.TYPE_CHECKING`` guard. I am proposing the same for ``NDimArray``.

Reply:
> However, there is a change I would like to use in our codebase, to speed up `import sympy`.

When you say "our codebase" do you mean that you want to make changes to sympy or something else?

> `BasisDependent` is used in exactly the same way as `NDimArray`, and is protected by the `typing.TYPE_CHECKING` guard

The difference is that if I move the other import to top-level then I immediately get an ImportEror.

I'm not sure that I understand but it seems like you want to make other changes to sympy in future and moving this import would be needed as part of those changes. If that is the case then would it not make sense to just combine both changes in one PR?

Otherwise if you are not intending to make other changes in sympy itself then I don't understand how this would help solve an ImportError in any other codebase (or speed up importing sympy).

To be clear I don't object to this change: I'm just a bit confused by a PR that only changes this one thing.

Reply:
> When you say "our codebase" do you mean that you want to make changes to sympy or something else?

It is sympy, but it will be quite a large change so it is possible it will not be accepted. If it is not accepted we will fork sympy and make the change in our codebase. Of course, the preference is to merge it into sympy main branch so everyone benefits. Once, I validate it is working, I will create a discussion, seperately.

> I'm not sure that I understand but it seems like you want to make other changes to sympy in future and moving this import would be needed as part of those changes. 

Correct. 

> If that is the case then would it not make sense to just combine both changes in one PR?

Simply, this change is trivial, what I hope will follow is probably less so. I am used to an environment where small changes are preferred and large changes should be broken down. 

> I'm just a bit confused by a PR that only changes this one thing.

Understood, for the reason, stated above, smaller, simpler changes, and (in my opinion, this should be done anyway because of [PEP 484](https://peps.python.org/pep-0484/#runtime-or-type-checking), but I accept this is debatable).